### PR TITLE
Fix Issue 13650 - Cannot copy (w)char[] into (w)char[]

### PR DESCRIPTION
**Pull request by Poita**
The diff is a bit messy, but the change is essentially to remove the `isOutputRange!(Range2, ElementType!Range1)` condition from the template to allow `char[]` and `wchar[]` targets (they aren't output ranges). The check is now only at the end when we try to use the non-array path.

https://issues.dlang.org/show_bug.cgi?id=13650

### CODE REVIEW

**Comment by DmitryOlshansky** [comment]
2014... OMG and quite stinky bug btw.
This needs a rebase.  @Poita - still interested in reviving it?

**Comment by MetaLang** [comment]
I can revive this if @Poita doesn't respond in a few days.

**Comment by DmitryOlshansky** [comment]
@Poita when rabasing please nag reviewers, Github won't notify me on new commits/rebase.

**Comment by DmitryOlshansky** [comment]
After taking a good glance over it - I still do not see where the actual enabling change is? 95% of it is just moving stuff around which is okay but.. Any hints would help merge it.

**Comment by DmitryOlshansky** [comment]
> but the change is essentially to remove the isOutputRange!(Range2, ElementType!Range1) condition from the template to allow char[] and wchar[] targets (they aren't output ranges

See it now. Well, LGTM then.

**Comment by Poita** [comment]
Yes, that's the only change. The code movement wasn't by choice, it was just one of those cases where changing a condition had a cascading effect that either forced a refactor, or would result in duplicate or redundant code.

Sorry for not pinging on rebase, didn't realize there wasn't a notification.

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by CyberShadow** [comment]
This pull request introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=14817

File: std/algorithm/mutation.d [hunk 1]
```diff
@@ -295,58 +295,62 @@ See_Also:
     $(WEB sgi.com/tech/stl/_copy.html, STL's _copy)
  */
 TargetRange copy(SourceRange, TargetRange)(SourceRange source, TargetRange target)
-if (isInputRange!SourceRange && isOutputRange!(TargetRange, ElementType!SourceRange))
```
Comment (CyberShadow) [inline std/algorithm/mutation.d]: Removing the constraint here is probably what caused the regression.

**Pull request merged**